modules/oracle/marketObserve.go

```diff
@@ -0,0 +1,204 @@
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"time"
+	"vsc-node/modules/oracle/p2p"
+
+	"github.com/go-playground/validator/v10"
+)
+
+var (
+	v = validator.New(validator.WithRequiredStructEnabled())
+
+	// to be processed by the block producer, calculating median price
+	// and creating new block
+	broadcastPriceBuf = make([]p2p.AveragePricePoint, 0, 256)
+
+	// to be signed by the witness
+	newBlockBuf = make([]p2p.VSCBlock, 0, 256)
+)
+
+func (o *Oracle) marketObserve() {
+	pricePollTicker := time.NewTicker(priceOraclePollInterval)
+
+	for {
+		select {
+		case <-o.ctx.Done():
+			return
+
+		case <-pricePollTicker.C:
+			for _, api := range o.priceOracle.PriceAPIs {
+				go api.QueryMarketPrice(watchSymbols, o.observePriceChan)
+			}
+
+		case broadcastSignal := <-o.broadcastPriceSignal:
+			o.handleBroadcastSignal(broadcastSignal)
+			broadcastPriceBuf = broadcastPriceBuf[:0]
+			newBlockBuf = newBlockBuf[:0]
+			o.priceOracle.ResetPriceCache()
+
+		case newBlock := <-o.broadcastPriceBlockChan:
+			// TODO: move this channel to witness processing
+			newBlockBuf = append(newBlockBuf, newBlock)
+
+		case avgPricePoints := <-o.broadcastPriceChan:
+			now := time.Now().UTC().Unix()
+			for i := range avgPricePoints {
+				avgPricePoints[i].UnixTimeStamp = now
+			}
+			broadcastPriceBuf = append(broadcastPriceBuf, avgPricePoints...)
+
+		case msg := <-o.msgChan:
+			// TODO: which one to send broadcast within the network?
+			if err := o.service.Send(msg); err != nil {
+				log.Println("[oracle] failed broadcast message", msg, err)
+				continue
+			}
+
+			jbytes, err := json.Marshal(msg)
+			if err != nil {
+				log.Println("[oracle] failed serialize message", msg, err)
+				continue
+			}
+			o.p2pServer.SendToAll(p2p.OracleTopic, jbytes)
+
+		case pricePoints := <-o.observePriceChan:
+			o.priceOracle.ObservePricePoint(pricePoints)
+
+			/*
+				case btcHeadBlock := <-o.blockRelayChan:
+					fmt.Println("TODO: validate btcHeadBlock", btcHeadBlock)
+
+				case blockRelaySignal := <-o.blockRelaySignal:
+					fmt.Println(blockRelaySignal)
+					go o.relayBtcHeadBlock()
+			*/
+
+		}
+	}
+}
+
+func (o *Oracle) handleBroadcastSignal(sig blockTickSignal) {
+	// local price
+	medianPriceBuf := make([]p2p.AveragePricePoint, 0, len(watchSymbols))
+
+	for _, symbol := range watchSymbols {
+		avgPricePoint, err := o.priceOracle.GetAveragePrice(symbol)
+		if err != nil {
+			log.Println("symbol not found in map:", symbol)
+			continue
+		}
+		medianPriceBuf = append(medianPriceBuf, *avgPricePoint)
+	}
+
+	o.msgChan <- &p2p.OracleMessage{
+		Type: p2p.MsgPriceOracleBroadcast,
+		Data: medianPriceBuf,
+	}
+
+	if !sig.isWitness && !sig.isBlockProducer {
+		return
+	}
+
+	if sig.isBlockProducer {
+		ts := time.Now().UTC().Unix()
+		for i := range medianPriceBuf {
+			medianPriceBuf[i].UnixTimeStamp = ts
+		}
+
+		// room for network latency
+		medianPricePoints := makeMedianPrices(medianPriceBuf)
+		nodeId := o.conf.Get()
+		vscBlock, err := p2p.MakeVscBlock(
```

Comment:
I am a bit confused what you mean by vscBlock here. We aren't signing new vsc blocks within this process so I'd be curious what the naming means here.

Reply:
I think a better name might have been `OracleBlock`? the idea is to store the median prices + volumes in this block and broadcast it within the network for peers to verify and sign